### ISSUE TITLE
Add a jump action

### ISSUE THREAD

**Issue by jrslagle** [Master-Foo/game#5]
Add a jump action with the spacebar key. 

I'm also making a list of all the changes required for adding any action. That gives us something to follow when adding more actions in the future.

**Comment by jrslagle** [comment]
General question: do you want us to pull request to your master branch?

**Comment by jrslagle** [comment]
I claim this feature by the way. I'm already about half way into it.

**Comment by Master-Foo** [comment]
I just added issue: Add a hold_item action to Avatar #6

We should probably work on that before adding a jump action.

I'm not saying we DON'T make a jump action. It's just that at this juncture, jumping doesn't really accomplish something useful. Meanwhile holding an object DOES do something useful so it should be prioritized over jumping.

FYI: When you are done, do a pull request. I'll check out your code and test it and make some suggestions. If it passes, I'll merge it. If it doesn't, I'll deny and give my reasons.

**Comment by jrslagle** [comment]
I could likewise argue that jumping is immediately usable, while holding does nothing until there are items to hold. We'll get to both in time. I'll write holding down. But I'd rather not reverse the ~20 modifications I've already made.

**Comment by Master-Foo** [comment]
What I mean is, jumping doesn't really accomplish anything. What are we jumping over? I'm not saying we don't need it. We just have other priorities. But if you really want to do this. I'm not going to twist your arm. Go for it.

**Comment by Master-Foo** [comment]
BTW, not related to holding / jumping. I just did a fairly major code cleanup. You might want to update my changes to your code base.

**Comment by jrslagle** [comment]
Just ran into a snag - jumping lasts for a certain amount of time and then ends. And jumping should probably disable walking and running.

Holding could end or not when letting go of the hold key.

**Comment by jrslagle** [comment]
K. I'll rebase before pushing.

**Comment by Master-Foo** [comment]
You will need to implement some jumping physics noteably there will need to be a z velocity that determines how high you jump. Which probably means we should revamp the current velocity mechanism in the Entity. Which is probably should be a prerequisite issue to this one. In fact... I'm going to make that issue now. You add that issue to the OP in this issue and we'll come back to this issue when that is accomplished.

Then, we are going to have to do some "cheating" in the rendering. Because there isn't a z axis on the 2d world. So, we'll have to create the illusion of jumping by moving the character on the y axis instead of the non-existant z-axis in the 2d world.

**Comment by Master-Foo** [comment]
I just added issue: Revamp velocity system. #8

**Comment by Master-Foo** [comment]
FYI: Relevant to discussion. The newest revamp created a new class called AnimateEntity which now holds the logic for velocity. Things that move (like Avatar) inherit from AnimateEntity. "Dead" things, (like Tile) will inherit from InanimateEntity (Which is yet to be implemented)

**Comment by Master-Foo** [comment]
Also FYI, abstract classes (classes which are always inherited), like Entity and AnimateEntity are now in their own folder called /source/abstract

**Comment by jrslagle** [comment]
this merge is going to be hell. lol

**Comment by Master-Foo** [comment]
This merge will either be smooth, or a valuable learning experience for you.

It occurs to me that you can probably use one of the "Run" graphics as a "jump" graphic. The one where the guy is like "Look at me I'm stretching my legs out really far"

**Comment by jrslagle** [comment]
Yea, I just kept frames 0, 1, and 2 and put the interval at 0.2 seconds each.

**Comment by jrslagle** [comment]
K. I'm getting stuck trying to get MoveState from EntityModel to be accessible by AvatarView. You had been basing which animation to display by looking at the velocity. I want to use MoveState, but those constants are inside another class in EntityModel. How do I get at them??

**Comment by Master-Foo** [comment]
I'll take a look at it tomorrow. I'm going to make one last commit tonight, which is adding a stupid rock to the game. Then I'm calling it quits for the night.

Push your code to your repository so I can see what you've done and I'll pick it up in the morningish

**Comment by jrslagle** [comment]
Dam, I went to bed without pushing. I won't be home til late tonight. I haven't done a rebase with your reorganized code yet either, so you'd probably hate what you see anyway. Lol.

**Comment by jrslagle** [comment]
On passing state from model to view, there was a solution to this I read in a textbook on Android development. It seemed unnecessarily convoluted at the time and was hard to follow, but that's my next best thing to try.

**Comment by jrslagle** [comment]
The more I think about this, the more I think we'll avoid future bugs by avoiding special move states. Otherwise you'll run into bugs like what happens when multiple actions are happening to you at once. I motion to refactor for a force diagram.

Net force on you is (gravity + normal force + friction). Anything that strikes you just gets appended to a force array and you go flying. 

Thoughts?

**Comment by Master-Foo** [comment]
push your code and I'll take a look at it. I'm really not sure what you are talking about "passing state from model to view". view has normal and direct access to the model's data members. Just use the model state as if it were a member of the view.

**Comment by jrslagle** [comment]
Actually, I'm spending so much work trying to refactor the code to get jumping to be possible that I think I'm going to rebase with your code first and that should save me refactoring twice. 

1. Rebase with Mower's refactoring. 
2. Change avatar animations to be based on move state (and not an elif chain of velocities). 
3. Give each action its own velocity in avatar model. 
4. Get jumping to work.

**Comment by Master-Foo** [comment]
Tackle the Revamp velocity system. #8 issue first and come back to this.

**Comment by Master-Foo** [comment]
I recently did some cleanup in Avatar. As a result, making new actions is super simple.

Check out the [commit](https://github.com/Master-Foo/game/commit/08394514628ebda9d42a3cc9eef924c051877aa2) for issue: Create a Mining action for Human #39 as an example. It's literally only 15 lines of code. Also, it demonstrates nicely where code should go as far as Controller, Model, View as well as Where it should go in the hierarchy of inheritance.

For instance the events listening for key press are in **Avatar** controller. The logic is in **Human** model.

Adding a jump action will be a little tougher because you are going to need to implement gravity and give some attention to the Avatar's z axis, which exists, but has been neglected. But, regardless, doing this issue should be much simpler.

I'm going to make gravity a new prerequisite to this issue. That should be a separate project and it will affect all entities, not just Avatar.

**Comment by Master-Foo** [comment]
I just added issue: Add Gravity #49
